File: Sources/ChatUnpackCore/Domain/ScanProgress.swift
import Foundation

public enum ScanPhase: String, Sendable, Codable {
  case movingToTop
  case capturing
  case recognizing
  case assembling
  case scrolling
  case restoringPosition
  case paused
  case completed
  case incomplete

  public var displayName: String {
    switch self {
    case .movingToTop:
      return "回到顶部"
    case .capturing:
      return "正在捕获"
    case .recognizing:
      return "正在识别"
    case .assembling:
      return "正在拼接"
    case .scrolling:
      return "正在滚动"
    case .restoringPosition:
      return "正在恢复位置"
    case .paused:
      return "已暂停"
    case .completed:
      return "已完成"
    case .incomplete:
      return "提取未完成"
    }
  }
}

public struct ScanProgress: Sendable, Equatable {
  public var phase: ScanPhase
  public var viewportCount: Int
  public var messageCount: Int
  public var lowConfidenceCount: Int
  public var percent: Double?
  public var reason: String?

  public init(
    phase: ScanPhase,
    viewportCount: Int = 0,
    messageCount: Int = 0,
    lowConfidenceCount: Int = 0,
    percent: Double? = nil,
    reason: String? = nil
  ) {
    self.phase = phase
    self.viewportCount = viewportCount
    self.messageCount = messageCount
    self.lowConfidenceCount = lowConfidenceCount
    self.percent = percent
    self.reason = reason
  }
}

File: windows/src/ChatUnpack.Core/Domain/ScanProgress.cs
namespace ChatUnpack.Core.Domain;

public enum ScanPhase
{
  MovingToTop,
  Capturing,
  Recognizing,
  Assembling,
  Scrolling,
  RestoringPosition,
  Paused,
  Completed,
  Incomplete
}

public static class ScanPhaseNames
{
  public static string DisplayName(this ScanPhase phase)
  {
    return phase switch
    {
      ScanPhase.MovingToTop => "回到顶部",
      ScanPhase.Capturing => "正在捕获",
      ScanPhase.Recognizing => "正在识别",
      ScanPhase.Assembling => "正在拼接",
      ScanPhase.Scrolling => "正在滚动",
      ScanPhase.RestoringPosition => "正在恢复位置",
      ScanPhase.Paused => "已暂停",
      ScanPhase.Completed => "已完成",
      ScanPhase.Incomplete => "提取未完成",
      _ => string.Empty
    };
  }
}

public sealed class ScanProgress : IEquatable<ScanProgress>
{
  public ScanProgress(
    ScanPhase phase,
    int viewportCount = 0,
    int messageCount = 0,
    int lowConfidenceCount = 0,
    double? percent = null,
    string? reason = null)
  {
    Phase = phase;
    ViewportCount = viewportCount;
    MessageCount = messageCount;
    LowConfidenceCount = lowConfidenceCount;
    Percent = percent;
    Reason = reason;
  }

  public ScanPhase Phase { get; set; }
  public int ViewportCount { get; set; }
  public int MessageCount { get; set; }
  public int LowConfidenceCount { get; set; }
  public double? Percent { get; set; }
  public string? Reason { get; set; }

  public bool Equals(ScanProgress? other)
  {
    return other is not null
      && Phase == other.Phase
      && ViewportCount == other.ViewportCount
      && MessageCount == other.MessageCount
      && LowConfidenceCount == other.LowConfidenceCount
      && Percent == other.Percent
      && Reason == other.Reason;
  }

  public override bool Equals(object? obj)
  {
    return Equals(obj as ScanProgress);
  }

  public override int GetHashCode()
  {
    return HashCode.Combine(
      Phase,
      ViewportCount,
      MessageCount,
      LowConfidenceCount,
      Percent,
      Reason);
  }
}
